pkg/reconciler/nodeconfig_reconciler.go

```diff
@@ -61,95 +68,208 @@ func (ncr *NodeConfigReconciler) reconcileDebounced(ctx context.Context) error {
+	nodeConfigs, err := ncr.listConfigs(ctx)
+	if err != nil {
+		return fmt.Errorf("error listing configs: %w", err)
+	}
+
+	available := len(nodes)
+	ready := 0
+	ongoing := 0
+	shouldExit, err := ncr.processConfigs(ctx, nodeConfigs.Items, revisionToDeploy, &ready, &ongoing)
+	if err != nil {
+		return fmt.Errorf("error processing configs: %w", err)
+	}
+	if shouldExit {
+		return nil
+	}
+
+	nodesToDeploy := getNodesToDeploy(nodes, nodeConfigs.Items, revisionToDeploy)
+
+	if err := ncr.updateRevisionCounters(ctx, revisionToDeploy, available, ready, ongoing); err != nil {
+		return fmt.Errorf("error updating revision %s counters: %w", revisionToDeploy.Name, err)
 	}
 
-	revisionToDeploy.Status.IsInvalid = false
-	if err := ncr.client.Status().Update(ctx, revisionToDeploy); err != nil {
-		return fmt.Errorf("error validating revision %s: %w", revisionToDeploy.Name, err)
+	if revisionToDeploy.Status.Ongoing < ncr.updateLimit && len(nodesToDeploy) > 0 {
+		if err := ncr.deployNodeConfig(ctx, nodesToDeploy[0], revisionToDeploy); err != nil {
+			return fmt.Errorf("error deploying node configurations: %w", err)
+		}
 	}
 
 	// remove all but last known valid revision
-	if err := ncr.revisionCleanup(ctx, revisionToDeploy); err != nil {
+	if err := ncr.revisionCleanup(ctx); err != nil {
 		return fmt.Errorf("error cleaning redundant revisions: %w", err)
 	}
 
 	return nil
 }
 
-func (ncr *NodeConfigReconciler) revisionCleanup(ctx context.Context, validRevision *v1alpha1.NetworkConfigRevision) error {
-	revisions, err := listRevisions(ctx, ncr.client)
-	if err != nil {
-		return fmt.Errorf("error listing revisions: %w", err)
+func getFirstValidRevision(revisions []v1alpha1.NetworkConfigRevision) *v1alpha1.NetworkConfigRevision {
+	var result *v1alpha1.NetworkConfigRevision
+	for i := range revisions {
+		if revisions[i].Status.IsInvalid {
+			continue
+		}
+		result = &revisions[i]
+		break
 	}
+	return result
+}
 
-	// remove all but last known valid revision
-	for i := range revisions.Items {
-		if revisions.Items[i].Spec.Revision != validRevision.Spec.Revision {
-			if err := ncr.client.Delete(ctx, &revisions.Items[i]); err != nil {
-				if !apierrors.IsNotFound(err) {
-					return fmt.Errorf("error deleting revision %s: %w", revisions.Items[i].Name, err)
+func (ncr *NodeConfigReconciler) processConfigs(ctx context.Context, configs []v1alpha1.NodeNetworkConfig, revision *v1alpha1.NetworkConfigRevision, ready, ongoing *int) (bool, error) {
+	for i := range configs {
+		// Every NodeNetworkConfig obejct should have 2 owner references - for NodeConfigRevision and for the Node. If there is only one owner reference,
+		// it means that either node or revision were deleted, so the config itself can be deleted as well.
+		if len(configs[i].ObjectMeta.OwnerReferences) < numOfRefs {
+			if err := ncr.client.Delete(ctx, &configs[i]); err != nil {
+				return true, fmt.Errorf("error deleting redundant node config - %s: %w", configs[i].Name, err)
```

Comment:
usually a good idea to ignore not found errors when deleting. Can happen here as a race condition